Session12

Hyperbolic4183 opened the pull request:
## 行ったこと(8/11)
- SPMによるエラーに対して適切なハンドリングが行われるようにした
- フォアグランドに戻った時にクラッシュするバグを直した
- 複数の非同期処理が終わった後にインジケータが終了するように修正した。
- エラーが起きた時にインジケータが周り続けるバグを直した。

## 懸念点
- showIndicator(while processings) は時間に揺らぎのある複数の処理を同時におこない、処理が全て終了した時にインジケータのアニメーションをストップする関数であるが、このような関数のベストプラクティスがあれば知りたい。

Hyperbolic4183 committed:
fixes #5
ネストを解消し、エラーハンドリングが適切に行われるようにした。

Hyperbolic4183 committed:
close #3
関数を分けた。

Hyperbolic4183 committed:
close #7
二つのfetch関数に時間的な揺らぎがあった場合にも対応できるようにした。

Hyperbolic4183 committed:
行間を調整

ykws commented on Example/Example/UI/Weather/Base.lproj/Weather.storyboard:
```diff
@@ -1,10 +1,12 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="16096" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="BYZ-38-t0r">
```
![ask-badge](https://img.shields.io/badge/review-ask-blue.svg)

UI に不具合はありましたか？

Hyperbolic4183 replied:
ありません。
変えたつもりなかったのにコミットしてしまってました(*_*)

ykws replied:
不具合の修正と関係のない変更は元に戻してみてください。

ykws commented on Example/Example/Model/WeatherModel.swift:
```diff
@@ -41,17 +41,25 @@ class WeatherModelImpl: WeatherModel {
+                assertionFailure("予期せぬエラーが発生した")
             }
+        } catch {
+            completion(.failure(.invalidParameterError))
```
![ask-badge](https://img.shields.io/badge/review-ask-blue.svg)

`.invalidParameterError` を `WeatherError` の `switch` と分けているのは何か理由がありますか？

Hyperbolic4183 replied:
これdo{}内の
```
let requestJson = try jsonString(from: request)
let responseJson = try YumemiWeather.syncFetchWeather(requestJson)
let response = try self.response(from: responseJson)
```
で吐かれるエラーの属するenumが異なっているので、分けざるを得なくなってしまった感じです。
こういう時、どういう対応をすべきでしょうか？

ykws replied:
YumemiWeather.syncFetchWeather でエラーとなった場合、常に WeatherError の invalidParameterError として扱うのが適切でしょうか？

ykws commented on Example/Example/UI/Weather/WeatherViewController.swift:
```diff
@@ -30,7 +30,7 @@ class WeatherViewController: UIViewController {
         
         NotificationCenter.default.addObserver(forName: UIApplication.didBecomeActiveNotification, object: nil, queue: nil) { [unowned self] notification in
-            self.loadWeather(notification.object)
+            loadWeather()
```
![ask-badge](https://img.shields.io/badge/review-ask-blue.svg)

こう変更できてしまうので、そもそもブロック処理を指定する Observer の登録が不自然に見えるので、別の登録方法で実現できそうか調べてみてください。

ykws reviewed (changes requested):
### Summary

不具合4件中2件発見 🎉 

1. SPMによるエラーに対して適切なハンドリングが行われるようにした 🤔 
2. フォアグランドに戻った時にクラッシュするバグを直した 👍  
3. 複数の非同期処理が終わった後にインジケータが終了するように修正した 🙅 
4. エラーが起きた時にインジケータが周り続けるバグを直した 👍  

---

1. SPM とは何を指していますでしょうか？
2. 不具合の指摘はばっちりです！　問題も解消されています。より適切な変更方法があるので、インラインコメントを参考に調べてみてください。
3. 災害情報の表示は非同期処理ではないので、 **複数の非同期処理** とはなっていないです。よって不具合とはカウントしません。
4. 不具合の指摘はばっちりです！　問題も解消されています。エラーの扱いが適切かどうか確認をお願いします。

不具合としてはあと2つあります。

---

それとコミットメッセージに関して気になったので補足です。

- 1行目が対応内容の要約
- 2行目は空白、
- 3行目以降に詳細及び、 Issue 番号と対応状況（ refs or fixes ）

となっていると見やすいです。

が、プロジェクトによっては 1行目に Issue 番号を prefix とするルールもあるので、
リポジトリで統一できていれば良いです。

ただし、プロジェクトでは Issue なしの変更は認めていないこともあるのでこのルールが適用しやすいですが、
Issue なしの変更も許容すると、 1行目に Issue とするのは、一貫性を保ちづらいので、
3行目以降の詳細に含めるのが無難です。今回の PR でも混ざっていたので気になりました。

Hyperbolic4183 commented:
@ykws 
SPMは外部SwiftPackgeManagerのつもりでした。もしかしたら4とかぶっているかもしれません🙇‍♂️

ykws commented:
> SPMは外部SwiftPackgeManagerのつもりでした

ですよね。 SPM の差分がコミットにないようだったので、コミット漏れでしょうか？

Hyperbolic4183 committed:
エラー処理を見直した

Hyperbolic4183 committed:
revert

誤って変更してしまった。

Hyperbolic4183 committed:
fix observerへの登録方法の変更

https://github.com/Hyperbolic4183/YumemiWeather-session12/pull/8#discussion_r686926050

Hyperbolic4183 committed:
update UIButtonとUIActivityIndicatorViewを拡張

UIButtonのisEnabledを切り替えるメソッドとUIActivityIndicatorのアニメーションを切り替えるメソッドを拡張した。refs #9

Hyperbolic4183 committed:
update UIButtonとViewControllerを紐付けた

UIButtonのisEnabledにアクセスするために紐付けを行なった。refs #9

Hyperbolic4183 committed:
fix reload中にボタンが非活性になるようにした。

fixes #9

Hyperbolic4183 committed:
fix エラーが起きた時にクラッシュするバグを修正した

Hyperbolic4183 committed:
update ライブラリの機能拡張

DI可能になるように外部ライブラリR.Swiftを拡張した。 refs #11

Hyperbolic4183 committed:
update モデルを外部から注入できるようにした

イニシャライザにWeatherModelとDisasterModelが必要になるようにした。 refs #11

Hyperbolic4183 committed:
update WeatherViewControllerのイニシャライザの方法を変更した

refs #11

Hyperbolic4183 committed:
fix 壊れていたモックを修正した。

モックをテストしたい要素によってイニシャライズできるようにし、fetchWeather()を定義した。refs #10

Hyperbolic4183 committed:
fix ViewControllerからUI部品を取り出す関数を定義した。

UIStackViewとしてはたらいているUIViewを取り出す関数を定義し、その中に埋め込まれたUILabel,UIImageViewを取り出す。 refs #10

Hyperbolic4183 committed:
fix 壊れていたテストを修正した
refs #10

Hyperbolic4183 committed:
update storyboardに貼り付けられたUI部品をコード内で扱えるようにした。

identifierを指定し、accessibilityを変更した。 refs #10

Hyperbolic4183 committed:
update エラーが起きた時に画像を設定する関数を定義

エラーが起きたときにweatherImageView.imageに使用する関数を定義 refs #12

Hyperbolic4183 committed:
fix エラー発生時の処理を変更した

エラーが発生した時にViewControllerを閉じるのではなく、専用の画像と文字列を設定するようにした。 refs #12

Hyperbolic4183 committed:
update reloadButtonにidを追加

refs #10

Hyperbolic4183 committed:
update UI部品を抜き出す関数を修正した

返り値のオプショナルを外し、lazyを使い処理を早くした。 refs #10

Hyperbolic4183 committed:
update エラーを伝達するタイミングを変更した

Hyperbolic4183 committed:
move プロトコルの定義の場所を変更した

Hyperbolic4183 committed:
update エラーハンドリングを見直した。

Hyperbolic4183 committed:
update アクセス修飾子を追加した

Hyperbolic4183 committed:
update イニシャライザを変更した

Hyperbolic4183 committed:
update モデルモックを作成

refs #10

Hyperbolic4183 committed:
update テストを変更した

refs #10

Hyperbolic4183 committed:
update Viewの更新の責務を切り分け、テスト可能にした。

Hyperbolic4183 committed:
update WeatherViewControllerの単体テストを追加した

WeatherHandlerを呼ぶテストを追加した refs #10

Hyperbolic4183 committed:
update キューを切り替えるための構造体を作成した

refs #10

Hyperbolic4183 committed:
update 画面の更新を行うクラスのモックを作成した。

refs #10

Hyperbolic4183 committed:
update 画面の更新を行うクラスのキューの指定をWeatherViewControllerで行うようにし、テスト時に切り替えられるようにした。

refs #10

Hyperbolic4183 committed:
fix WeatherViewControllerの結合テストとして成立させた

refs #10

Hyperbolic4183 committed:
update キューの切り替えに引数ラベルか追加できるようにした。

Hyperbolic4183 committed:
update WeatherHandle.handleに引数をつけた

Hyperbolic4183 commented:
# 行ったこと(9/17)
1 フォアグランドに戻った時にクラッシュするバグを直した(前回確認済み)
2 エラーが起きた時にインジケータが周り続けるバグを直した(前回確認済み)
3 壊れていたテストを直した #10  <- New
4 エラーが起きた時に画面がクローズするバグを直した #12 <- New
5 ViewControllerをDI可能にした #11 <- New
6 fetch中にボタンが押せるバグを修正した #9 <- New

## 3
### 実装の概要
テストが壊れていたので、WeatherViewControllerとviewの結合テストを書き直した。
WeatherViewControllerの単体テストも追加した。

### 詳細
WeatherViewControllerでは、複数の処理が全て終わるまでインジケータを表示し続ける'showIndicator(processings: Void...)'を実装している。また、天気の画像の切り替えやエラーアラートの表示といった画面の更新を'WeatherHandler'構造体の'handle(from )'によって行っている。
'showIndicator(processings: Void...)'の内部でglobalQueueへの切り替えを、'WeatherHandler.handler'を呼び出す時にmainQueueへの切り替えを行っている。そのため、テストの際にはそれらを考慮する必要があるため、DI可能にした。

## 4
### 実装の概要
エラーが起きてアラートのOKボタンが押された後、画面を閉じるのではなく、エラーが起きた時に画像を差し替えるようにした。

### 詳細
天気の情報の取得はエラーが起きるが、災害の情報の取得はエラーが起きない。しかし、天気のエラーが起きた時にViewがdismissされてしまうため、本来正常に表示されるはずの災害の情報が見れなくなっている。
この問題を解決するために、天気の情報の取得の失敗時には画像をエラー画像に差し替えるように変更した。

## 5
### 実装の概要
R.Swiftを拡張して、DI可能にし、テストの時にモックを注入できるようにした。

## 6
### 実装の概要
fetch中のボタンを非活性にした。

Hyperbolic4183 committed:
fix 細かな修正

final private 修飾子をつけ、コメントを書いた

novr commented on Example/Example/UI/Weather/QueueScheduler.swift:
```diff
@@ -0,0 +1,49 @@
+extension QueueScheduler {
+    static let main = Self(
+        schedule: { group, action in
+            DispatchQueue.main.async(group: group) {
```
[MUST] DispatchGroupはnil不可です
> group: The dispatch group to associate with the work item. This parameter cannot be NULL.
https://developer.apple.com/documentation/dispatch/dispatchqueue/2300095-async

novr replied:
async(group:qos:flags:execute:)はnil許容だった

novr commented on Example/Example/UI/Root/RootViewController.swift:
```diff
@@ -8,29 +8,40 @@
-        // Pass the selected object to the new view controller.
+//MARK:- RswiftでstoryboardでDIができるように拡張する
+import Rswift
+public extension StoryboardResourceWithInitialControllerType {
```
[IMO] RootViewControllerではなく共通の拡張なので別ファイルで定義しましょう

novr commented on Example/Example/UI/Weather/WeatherViewController.swift:
```diff
@@ -8,94 +8,109 @@
+        NotificationCenter.default.addObserver(self, selector: #selector(reload), name: UIApplication.didBecomeActiveNotification, object: nil)
+    }
+    
+    init?(coder: NSCoder, weatherModel: WeatherModel, disasterModel: DisasterModel = DisasterModelImpl(), weatherHandler: WeatherHandler = .live, mainQueueScheduler: QueueScheduler = .main, globalQueueScheduler: QueueScheduler = .global) {
```
[NITS] 長くなりすぎる場合は適宜改行など
```suggestion
    init?(coder: NSCoder,
    	  weatherModel: WeatherModel,
    	  disasterModel: DisasterModel = DisasterModelImpl(),
    	  weatherHandler: WeatherHandler = .live,
    	  mainQueueScheduler: QueueScheduler = .main,
    	  globalQueueScheduler: QueueScheduler = .global) {
```

novr commented on Example/Example/UI/Weather/WeatherViewController.swift:
```diff
@@ -8,94 +8,109 @@
+
+private extension UIButton {
+    func toggleEnabled() {
+        self.isEnabled = !self.isEnabled
```
[NITS] use toggle()
```suggestion
        self.isEnabled.toggle()
```
https://developer.apple.com/documentation/swift/bool/2994863-toggle

novr commented on Example/Example/UI/Weather/WeatherViewController.swift:
```diff
@@ -8,94 +8,109 @@
+    
+    private func switchView() {
+        activityIndicator.toggleAnimation()
+        reloadButton.toggleEnabled()
```
[NITS] こちらでも十分意味は伝わるかと
```suggestion
        reloadButton.isEnabled.toggle()
```

Hyperbolic4183 committed:
fix Rswiftの拡張を別ファイルで行うようにした。

refs #8

Hyperbolic4183 committed:
fix イニシャライザで改行するようにした

refs #8

Hyperbolic4183 committed:
fix toggle()を用いて反転するようにした。

refs #8

novr reviewed (approved):
![LGTM](https://image.lgtmoon.dev/118228)
